OpenCL/m24000-pure.cl

```diff
@@ -0,0 +1,881 @@
+    w2[1] = tmp.y;
+    w2[2] = tmp.z;
+    w2[3] = tmp.w;
+
+    tmp = tmps[gid].P[l + 3];
+
+    w3[0] = tmp.x;
+    w3[1] = tmp.y;
+    w3[2] = tmp.z;
+    w3[3] = tmp.w;
+
+    sha256_hmac_update_64 (&ctx, w0, w1, w2, w3, 64);
+  }
+
+  w0[0] = 1;
+  w0[1] = 0;
+  w0[2] = 0;
+  w0[3] = 0;
+  w1[0] = 0;
+  w1[1] = 0;
+  w1[2] = 0;
+  w1[3] = 0;
+  w2[0] = 0;
+  w2[1] = 0;
+  w2[2] = 0;
+  w2[3] = 0;
+  w3[0] = 0;
+  w3[1] = 0;
+  w3[2] = 0;
+  w3[3] = 0;
+
+  sha256_hmac_update_64 (&ctx, w0, w1, w2, w3, 4);
+
+  sha256_hmac_final (&ctx);
+
+  char version = esalt_bufs[DIGESTS_OFFSET].version;
+
+  u32 iv[4] = { 0 };
+
+  u32 res[20]; // full would be 24 x u32 (96 bytes)
+
+  if (version == '8')
+  {
+
+    #define KEYLEN 60
+
+    u32 ks[KEYLEN];
+
+    AES256_set_decrypt_key (ks, ctx.opad.h, s_te0, s_te1, s_te2, s_te3, s_td0, s_td1, s_td2, s_td3);
+
+    for (u32 i = 0; i < 20; i += 4) // 96 bytes output would contain the full 32 byte checksum
+    {
+      u32 data[4];
+
+      data[0] = esalt_bufs[DIGESTS_OFFSET].ciphertext[i + 0];
+      data[1] = esalt_bufs[DIGESTS_OFFSET].ciphertext[i + 1];
+      data[2] = esalt_bufs[DIGESTS_OFFSET].ciphertext[i + 2];
+      data[3] = esalt_bufs[DIGESTS_OFFSET].ciphertext[i + 3];
+
+      u32 out[4];
+
+      aes256_decrypt (ks, data, out, s_td0, s_td1, s_td2, s_td3, s_td4);
+
+      res[i + 0] = hc_swap32_S (out[0] ^ iv[0]);
+      res[i + 1] = hc_swap32_S (out[1] ^ iv[1]);
+      res[i + 2] = hc_swap32_S (out[2] ^ iv[2]);
+      res[i + 3] = hc_swap32_S (out[3] ^ iv[3]);
+
+      iv[0] = data[0];
+      iv[1] = data[1];
+      iv[2] = data[2];
+      iv[3] = data[3];
+    }
+  }
+
+
+  if (version == '9')
+  {
+    u32 sk[4];
+    u32 lk[40];
+
+    ctx.opad.h[0] = hc_swap32_S (ctx.opad.h[0]);
+    ctx.opad.h[1] = hc_swap32_S (ctx.opad.h[1]);
+    ctx.opad.h[2] = hc_swap32_S (ctx.opad.h[2]);
+    ctx.opad.h[3] = hc_swap32_S (ctx.opad.h[3]);
+    ctx.opad.h[4] = hc_swap32_S (ctx.opad.h[4]);
+    ctx.opad.h[5] = hc_swap32_S (ctx.opad.h[5]);
+    ctx.opad.h[6] = hc_swap32_S (ctx.opad.h[6]);
+    ctx.opad.h[7] = hc_swap32_S (ctx.opad.h[7]);
+
+    twofish256_set_key (sk, lk, ctx.opad.h);
+
+    for (u32 i = 0; i < 20; i += 4) // 96 bytes output would contain the full 32 byte checksum
+    {
+      u32 data[4];
+
+      data[0] = esalt_bufs[DIGESTS_OFFSET].ciphertext[i + 0];
+      data[1] = esalt_bufs[DIGESTS_OFFSET].ciphertext[i + 1];
+      data[2] = esalt_bufs[DIGESTS_OFFSET].ciphertext[i + 2];
+      data[3] = esalt_bufs[DIGESTS_OFFSET].ciphertext[i + 3];
+
+
+      u32 out[4];
+
+      twofish256_decrypt (sk, lk, data, out);
+
+      res[i + 0] = hc_swap32_S (out[0] ^ iv[0]);
+      res[i + 1] = hc_swap32_S (out[1] ^ iv[1]);
+      res[i + 2] = hc_swap32_S (out[2] ^ iv[2]);
+      res[i + 3] = hc_swap32_S (out[3] ^ iv[3]);
+
+      iv[0] = data[0];
+      iv[1] = data[1];
+      iv[2] = data[2];
+      iv[3] = data[3];
+    }
+  }
+
+  if (version == 'a')
+  {
+    printf("%s\n", "SERPENT");
```

Comment:
please remove this debugging code, no printf () should be in the production code